doc/sg_wr_mode.8
.TH SG_WR_MODE "8" "September 2023" "sg3_utils\-1.49" SG3_UTILS
.SH NAME
sg_wr_mode \- write (modify) SCSI mode page
.SH SYNOPSIS
.B sg_wr_mode
[\fI\-\-cfile=CF\fR] [\fI\-\-contents=H,H...\fR] [\fI\-\-dbd\fR]
[\fI\-\-force\fR] [\fI\-\-help\fR] [\fI\-\-len=10|6\fR]
[\fI\-\-mask=M,M...\fR] [\fI\-\-mfile=MF\fR] [\fI\-\-page=PG_H[,SPG_H]\fR]
[\fI\-\-raw\fR] [\fI\-\-rtd\fR] [\fI\-\-save\fR] [\fI\-\-six\fR]
[\fI\-\-verbose\fR] [\fI\-\-version\fR] \fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Writes a modified mode page to \fIDEVICE\fR. Uses the SCSI MODE SENSE (6
or 10 byte variant) command to fetch the existing mode data which includes
a mode page (or subpage). It then combines that with the contents,
potentially masked, and writes the modified mode page with the SCSI MODE
SELECT (6 or 10 byte variant) command. This utility does not modify
the block descriptor(s); if any block descriptors are fetched by the MODE
SENSE command then the same block descriptors are written back with the
following MODE SELECT command.
.PP
If the \fI\-\-rtd\fR option is given then most other options apart from
\fI\-\-save\fR, \fI\-\-len=\fR10|6\fR and \fI\-\-six\fR are ignored. In this
case only a MODE SELECT command is sent to the \fIDEVICE\fR with the RTD
bit (Revert To Defaults) set. This bit was added to this command in SPC\-5
revision 11, so older devices may not support it. The Extended Inquiry VPD
page has the RTD_SUP bit to indicate whether the \fIDEVICE\fR supports the
RTD bit in the MODE SELECT(6 and 10) commands. When the \fI\-\-rtd\fR option
is given the rest of this section can be ignored.
.PP
If a contents option is not given then the various components (i.e.
header, block descriptor(s) and mode page) of the "current" values of
the existing mode page are printed out. In this case the mode page is
not altered on the device.
.PP
If the contents are specified, and a mask is not specified, then the contents
must match the existing mode page in various aspects unless the
\fI\-\-force\fR option is given. These include length, mode page code and
subpage code if applicable. If all is well then the contents string is
written to \fIDEVICE\fR as the new mode page.
.PP
If both contents and mask strings are specified then only bit positions
in the contents corresponding to set bits in the mask are taken while the
existing mode page supplies bit positions corresponding to clear bits.
When a mask is given then the mask and/or the contents may be shorter
than the existing mode page. If the mask is shorter than the contents then
the remaining bytes are taken from the contents. If the contents are shorter
than the existing mode page then the remaining bytes are taken from the
existing mod page.
.PP
The force option allows the contents string to be written as the new
mode page without any prior checks on the existing mode page. This should
only be required for vendor specific mode pages. The existing mode data
is ignored apart from the block descriptors which can be suppressed with
the \fI\-\-dbd\fR option if need be.
.PP
Changing individual fields in a mode page is probably more easily done
with the sdparm utility. Fields can be identified by acronym or by a
numerical descriptor.
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-C\fR, \fB\-\-cfile\fR=\fICF\fR
where \fICF\fR is a file name containing the mode page. See the "HEX,
BINARY AND JSON FORMATS" section in the sg3_utils manpage for more
information. If the \fI\-\-raw\fR option is also given then the contents of
\fICF\fR are treated as binary.
.TP
\fB\-c\fR, \fB\-\-contents\fR=\fIH,H...\fR
where \fIH,H...\fR is a string of comma separated hex numbers each of
which should resolve to a byte value (i.e. 0 to ff inclusive). A (single)
space separated string of hex numbers is also allowed but the list needs to
be in quotes. This is the new contents of the mode page to be written to
\fIDEVICE\fR, potentially filtered by the mask string.
.TP
\fB\-c\fR, \fB\-\-contents\fR=\-
reads contents string from stdin. See the \fI\-\-cfile=CF\fR option above.
.TP
\fB\-d\fR, \fB\-\-dbd\fR
disable block descriptors (DBD flag in cdb). Some device types include
block descriptors in the mode data returned by a MODE SENSE command. If
so the same block descriptors are written by the MODE SELECT command.
This option instructs the MODE SENSE command not to return any block
descriptors. This would be a sensible default for this utility apart
from the fact that not all SCSI devices support the DBD bit in the cdb.
.TP
\fB\-f\fR, \fB\-\-force\fR
force the contents string to be taken as the new mode page, or at least
doesn't do checks on the existing mode page. Note that \fIDEVICE\fR may
still reject the new contents for the mode page. Cannot be given with
the \fI\-\-mask=M,M...\fR option.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-l\fR, \fB\-\-len\fR=10 | 6
length of the SCSI commands (cdb) sent to \fIDEVICE\fR. The default is 10
so 10 byte MODE SENSE and MODE SELECT commands are issued. Some old devices
don't support the 10 byte variants hence this option.
.TP
\fB\-m\fR, \fB\-\-mask\fR=\fIM,M...\fR
where \fIM,M...\fR is a string of comma separated hex numbers each of which
should resolve to a byte value (i.e. 0 to ff inclusive). A (single) space
separated string of hex numbers is also allowed but the list needs to be in
quotes. The mask chooses (bit by bit) whether the new mode page comes from
the contents (mask bit set) or from the existing mode page (mask bit clear).
If the mask string is shorter than the contents string then the remaining
bytes are taken from the contents string. If the contents string is shorter
than the existing mode page then the remaining bytes are taken from the
existing mode page (i.e. they are left unaltered).
.TP
\fB\-M\fR, \fB\-\-mfile\fR=\fIMF\fR
where \fIMF\fR is a file name containing the mode page mask. See the "HEX,
BINARY AND JSON FORMATS" section in the sg3_utils manpage for more
information. If the \fI\-\-raw\fR option is also given then the contents of
\fIMF\fR are treated as binary.
.TP
\fB\-p\fR, \fB\-\-page\fR=\fIPG_H\fR
where \fIPG_H\fR is the page code value to fetch and modify. The page code
is in hex and should be between 0 and 3e inclusive. Notice that page code
3f to fetch all mode pages is disallowed.
.TP
\fB\-p\fR, \fB\-\-page\fR=\fIPG_H,SPG_H\fR
where \fIPG_H\fR is the page code value and \fISPG_H\fR is the subpage code
value to fetch and modify. Both values are in hex. The subpage code should
be between 0 and fe inclusive. Notice that subpage code ff to fetch all
mode subpages (for a given mode page or all mode pages in the case of 3f,ff)
is disallowed.
.TP
\fB\-r\fR, \fB\-\-raw\fR
when this option is given then the file given by the \fI\-\-cfile=CF\fR
option is treated as binary (the default is to treat it as ASCII hexadecimal).
.TP
\fB\-R\fR, \fB\-\-rtd\fR
when this option is given most other actions are bypassed and a MODE
SELECT(6 or 10) command is sent to the \fIDEVICE\fR with the RTD bit set.
This will cause all current values (and saved values if the \fI\-\-save\fR
option is also given) of all mode pages to be reverted to their default
values.
.TP
\fB\-s\fR, \fB\-\-save\fR
changes the "saved" mode page when MODE SELECT is successful. By
default (i.e. when \fI\-\-save\fR is not used) only the "current" mode page
values are changed when MODE SELECT is successful. In this case the new mode
page will stay in effect until the device is reset (e.g.  power cycled).
When it restarts the "saved" values for the mode page will be re\-instated.
So to make changes permanent use the \fI\-\-save\fR option.
.br
When used with the \fI\-\-rtd\fR option then both the current and saved
values in each mode page are reverted to their default values. In the
absence of \fI\-\-save\fR option only the current values in each mode page
are reverted to their default values.
.TP
\fB\-6\fR, \fB\-\-six\fR
this option will cause the 6 byte variants of MODE SENSE and MODE SELECT
commands to be used. The default is to use the 10 byte options. This option
is equivalent to using the \fI\-\-len=6\fR option.

.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH NOTES
This utility does not check whether the contents string is trying to
modify parts of the mode page which are changeable. The device should
do that and if some part is not changeable then it should
report: "Invalid field in parameter list".
.PP
Some mode pages are not saveable. If so an attempt to use the \fI\-\-save\fR
option should cause an error to be reported from the device: "Illegal field
in cdb".
.PP
The device is required to do various checks before it accepts a new
mode page. If these checks fail then the mode page is not altered and
either a "parameter list length error" or an "invalid field in
parameter list" error is returned by the device in the sense data.
.PP
The recommended way to modify a mode page is to read it with a
MODE SENSE, modify some part of it then write it back to the
device with a MODE SELECT command. For example, reading an existing mode
page can be accomplished with 'sg_modes \-p=1a \-r /dev/sdb > mp_1a.txt' (the
power condition mode page). The mp_1a.txt file can be edited and then used
as the contents string to this
utility (e.g. 'sg_wr_mode \-p 1a \-s \-c \- /dev/sdb < mp_1a.txt').
.PP
Two fields differ between what is read from the device with MODE SENSE and
what is written to the device with MODE SELECT:
the mode data length is reserved (i.e. zero(es)) in a MODE
SELECT command while the PS bit ((sub)page byte 0 bit 7) in each
mode (sub)page is reserved (zero) in a MODE SELECT command.
The PS bit given in the contents string is zeroed unless
the \fI\-\-force\fR option is selected.
.SH EXAMPLES
This utility can be used together with the sg_modes utility. To re\-instate
the default mode page values (i.e. the mode page values chosen by the
manufacturer of the device) as both the current and saved mode page
values the following sequence could be used:
.PP
  $ sg_modes \-\-control=2 \-\-page=1a \-r /dev/sda > t


ChangeLog
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for pre-release sg3_utils-1.49 [20231027] [svn: r1060]
  - JSON: make output more consistent so most command
    responses have a *_paramter_data or similar sub-object
  - apply https://github.com/doug-gilbert/sg3_utils/pull/39
    and its revision [20230807] mainly for Android
  - sg_inq: update version descriptors to T10 table 20230814
  - sg_wr_mode: fix --contents and --cfile= handling
  - sg_vpd: inform about SPC-6 (sg_inq knew already)
    https://github.com/doug-gilbert/sg3_utils/issues/40
  - sg_inq+sg_vpd: show depopulation time in hours and
    minutes as well as its raw form: seconds
    - add RSCS field to Block limits ext. vpage [sbc5r5]
  - sg_turs: fix missing CDB on several codepaths
  - sg_requests: improve buffer truncation
  - sg_logs: correct tape Device statisitics lpage [0x14]
    serial number parameters [0x40 and 0x41] per
    https://github.com/doug-gilbert/sg3_utils/issues/45
  - sg_get_elem_status: various improvements
  - sg_stream_ctl: add --hex, --inhex=, --json= and --raw
    options; additions in sbc5r05
  - rescan-scsi-bus.sh: fix multipath resize without update 
    https://github.com/doug-gilbert/sg3_utils/pull/43.diff
    and 44.diff
  - new error code: SG_LIB_PROGRESS_NOT_READY for long
    duration 'in progress' additional sense codes
  - lib/sg_pt_solaris.c : fix compilation issue per
    https://github.com/doug-gilbert/sg3_utils/pull/41
    - add --mfile=MF option for decoding mask in the MF file
  - lib/sgj_haj_*() functions: tweak hex_haj argument meaning
  - scripts/00-scsi-sg3_config.rules : apply
    https://github.com/doug-gilbert/sg3_utils/pull/42
  - testing/sg_chk_inq_vd.c: test internal table against T10
    version descriptor file

Changelog for released sg3_utils-1.48 [20230801] [svn: r1042]
  - decoding utilities: add --json[=JO] and --js-file=JFN
    options, short form changed to -j[=JO]
  - sg_sat_datetime: new utility for setting or accessing
    the data and time on a ATA device using a SAT layer.
  - sg_z_act_query: new utility for sending either a
    Zone activate or Zone query command
  - sg_rep_density: new utility for decoding the response of
    Report density support command [ssc (tape)]
  - sg_rem_rest_elem: new utility for removing or restoring
    elements
  - sg_write_attr: new utility, Write Attribute is a SPC
    command that is used mainly by tape drives (ssc)
  - --hex option cleanup, especially in sg_inq and sg_vpd
    - swap meaning of '-H' and '-HH' in sg_ses
  - sg_rtpg: https://github.com/hreinecke/sg3_utils/pull/79
    applied
  - netbsd: experimental port, scsi only
  - rescan-scsi-bus.sh: with '-r' it crashed due to change in
    rev 867 in sg_inq.c from Device_type= to PDT= .
    Change script to use either
    - undo regression in rev 815 that added newline after
      each LUN in the default (no option) output
    - rev 815 changed the order of listing hosts from
      numeric to alphabetical, change it back to numeric
    - https://github.com/doug-gilbert/sg3_utils/pull/17
      applied with tweaks: add timeout parameter
    - clean $norm handling
    - fix handling of '-I <secs>' option
    - sgdevice26: do not traverse sg class if scsi_device
      is not added
    - add -no-lip-scan option
    - https://github.com/doug-gilbert/sg3_utils/pull/21
    - speed testonline() when peripheral_qualifier != 0
      see https://github.com/doug-gilbert/sg3_utils/issues/24
    - improve handling of USB SD card readers (with no cards
      present)
    - speed multipath scans with many LUNs, cache multipath
      LUN info in temporary file, see
      https://github.com/doug-gilbert/sg3_utils/issues/22
  - udev rules: restrict use of ambiguous device IDs patchset
    (3 part) from Martin Wilck applied
  - sg_rep_zones: add Report zone starting LBA granularity
    field in REPORT ZONES response [zbc2r12]
    - add --brief option, show part of header and last
      descriptor fetched
    - add --find ZT option to find the first occurrence of
      ZT; if ZT prefixed by - or ! find first not equal to ZT
    - add --statistics option
  - sg_get_elem_status: change '--maxlen=' option default to
    1056 (header plus 32 physical element status descriptors)
    - add current (maximum) number of depopulated elements
      fields
  - sg_get_lba_status: decode new LBA accessibility field
  - sg_decode_sense: add --nodecode option
  - sg_logs: tweak the meaning of --list option to more closely
    reflect the contents of log pages 0x0 and 0x0,0xff
    - make '-lll' set union of log pages 0x0 and 0x0,0xff
    - add --exclude option to exclude vendor specific pages
      and parameters
    - add --undefined option for hex format of undefined/
      unrecognized fields
    - for short binary fields, remove address (index) from
      the left hand side of each line of hex
    - improve 'last_n' log pages; supply VPD and mode pages

src/sg_wr_mode.c
/*
 * Copyright (c) 2004-2023 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <getopt.h>
#include <ctype.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_cmds_basic.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 * This program writes the given mode page contents to the corresponding
 * mode page on the given device.
 */

static const char * version_str = "1.32 20230901";

#define ME "sg_wr_mode: "

#define MX_ALLOC_LEN 2048
#define SHORT_ALLOC_LEN 252

#define EBUFF_SZ 256


static const struct option long_options[] = {
    {"cfile", required_argument, 0, 'C'},
    {"contents", required_argument, 0, 'c'},
    {"dbd", no_argument, 0, 'd'},
    {"force", no_argument, 0, 'f'},
    {"help", no_argument, 0, 'h'},
    {"len", required_argument, 0, 'l'},
    {"mask", required_argument, 0, 'm'},
    {"mfile", required_argument, 0, 'M'},
    {"page", required_argument, 0, 'p'},
    {"raw", no_argument, 0, 'r'},
    {"rtd", no_argument, 0, 'R'},
    {"save", no_argument, 0, 's'},
    {"six", no_argument, 0, '6'},
    {"verbose", no_argument, 0, 'v'},
    {"version", no_argument, 0, 'V'},
    {0, 0, 0, 0},
};


static void
usage()
{
    pr2serr("Usage: sg_wr_mode [--cfile=CF] [--contents=H,H...] [--dbd] "
            "[--force]\n"
            "                  [--help] [--len=10|6] [--mask=M,M...] "
            "[--mfile=MF]\n"
            "                  [--page=PG_H[,SPG_H]] [--raw] [--rtd] "
            "[--save]\n"
            "                  [--six] [--verbose] [--version] DEVICE\n"
            "  where:\n"
            "    --cfile=CF | -C CF    contents in a file called CF\n"
            "    --contents=H,H... | -c H,H...    comma separated string "
            "of hex numbers\n"
            "                                     that is mode page contents "
            "to write\n"
            "    --contents=- | -c -   read stdin for mode page contents"
            " to write\n"
            "    --dbd | -d            disable block descriptors (DBD bit"
            " in cdb)\n"
            "    --force | -f          force the contents to be written\n"
            "    --help | -h           print out usage message\n"
            "    --len=10|6 | -l 10|6    use 10 byte (def) or 6 byte "
            "variants of\n"
            "                            SCSI MODE SENSE/SELECT commands\n"
            "    --mask=M,M... | -m M,M...   comma separated "
            "string of hex\n"
            "                                numbers that mask contents"
            " to write\n"
            "    --mfile=MF | -M MF        mask in a file called MF\n"
            "    --page=PG_H | -p PG_H     page_code to be written (in hex)\n"
            "    --page=PG_H,SPG_H | -p PG_H,SPG_H    page and subpage code "
            "to be\n"
            "                                         written (in hex)\n"
            "    --raw | -r            contents of CF file decoded as "
            "binary\n"
            "    --rtd | -R            set RTD bit (revert to defaults) in "
            "cdb\n"
            "    --save | -s           set 'save page' (SP) bit; default "
            "don't so\n"
            "                          only 'current' values changed\n"
            "    --six | -6            do SCSI MODE SENSE/SELECT(6) "
            "commands\n"
            "    --verbose | -v        increase verbosity\n"
            "    --version | -V        print version string and exit\n\n"
            "writes given mode page with SCSI MODE SELECT (10 or 6) "
            "command\n");
}


/* Read hex numbers from command line or stdin. On the command line can
 * either be comma or space separated list. Space separated list need to be
 * quoted. For stdin (indicated by *inp=='-') there should be either
 * one entry per line, a comma separated list or space separated list.
 * Returns 0 if ok, or sg3_utils error code if error. */
static int
build_mode_page(const char * inp, bool is_file, bool as_binary,
                uint8_t * mp_arr, int * mp_arr_len, int max_arr_len)
{
    int in_len, k;
    unsigned int h;
    const char * lcp;
    char * cp;
    char * c2p;

    if ((NULL == inp) || (NULL == mp_arr) ||
        (NULL == mp_arr_len))
        return SG_LIB_LOGIC_ERROR;
    lcp = inp;
    in_len = strlen(inp);
    if (0 == in_len)
        *mp_arr_len = 0;
    if (is_file || ('-' == inp[0])) {
        return sg_f2hex_arr(inp, as_binary, false, mp_arr, mp_arr_len,
                            max_arr_len);
    } else {        /* hex string on command line */
        k = strspn(inp, "0123456789aAbBcCdDeEfF, ");
        if (in_len != k) {
            pr2serr("%s: error at pos %d\n", __func__, k + 1);
            return SG_LIB_SYNTAX_ERROR;
        }
        for (k = 0; k < max_arr_len; ++k) {
            if (1 == sscanf(lcp, "%x", &h)) {
                if (h > 0xff) {
                    pr2serr("%s: hex number larger than 0xff at pos %d\n",
                            __func__, (int)(lcp - inp + 1));
                    return SG_LIB_SYNTAX_ERROR;
                }
                mp_arr[k] = h;
                cp = (char *)strchr(lcp, ',');
                c2p = (char *)strchr(lcp, ' ');
                if (NULL == cp)
                    cp = c2p;
                if (NULL == cp)
                    break;
                if (c2p && (c2p < cp))
                    cp = c2p;
                lcp = cp + 1;
            } else {
                pr2serr("%s: error at pos %d\n", __func__,
                        (int)(lcp - inp + 1));
                return SG_LIB_SYNTAX_ERROR;
            }
        }
        *mp_arr_len = k + 1;
        if (k == max_arr_len) {
            pr2serr("%s: array length exceeded\n", __func__);
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    return 0;
}

/* Read hex numbers from command line (comma separated list).
 * Can also be (single) space separated list but needs to be quoted on the
 * command line. Returns 0 if ok, or 1 if error. */
static int
build_mask(const char * inp, bool is_file, bool as_binary,
           uint8_t * mask_arr, int * mask_arr_len, int max_arr_len)
{
    int in_len, k;
    unsigned int h;
    const char * lcp;
    char * cp;
    char * c2p;

    if ((NULL == inp) || (NULL == mask_arr) ||
        (NULL == mask_arr_len))
        return 1;
    lcp = inp;
    in_len = strlen(inp);
    if (0 == in_len)
        *mask_arr_len = 0;

    if (is_file || ('-' == inp[0])) {
        return sg_f2hex_arr(inp, as_binary, false, mask_arr, mask_arr_len,
                            max_arr_len);
    } else {        /* hex string on command line */
        k = strspn(inp, "0123456789aAbBcCdDeEfF, ");
        if (in_len != k) {
            pr2serr("%s: error at pos %d\n", __func__, k + 1);
            return 1;
        }
        for (k = 0; k < max_arr_len; ++k) {
            if (1 == sscanf(lcp, "%x", &h)) {
                if (h > 0xff) {
                    pr2serr("%s: hex number larger than 0xff at pos %d\n",
                            __func__, (int)(lcp - inp + 1));
                    return 1;
                }
                mask_arr[k] = h;
                cp = (char *)strchr(lcp, ',');
                c2p = (char *)strchr(lcp, ' ');
                if (NULL == cp)
                    cp = c2p;
                if (NULL == cp)
                    break;
                if (c2p && (c2p < cp))
                    cp = c2p;
                lcp = cp + 1;
            } else {
                pr2serr("%s: error at pos %d\n", __func__,
                        (int)(lcp - inp + 1));
                return 1;
            }
        }
        *mask_arr_len = k + 1;
        if (k == max_arr_len) {
            pr2serr("%s: array length exceeded\n", __func__);
            return 1;
        }
    }
    return 0;
}


int
main(int argc, char * argv[])
{
    bool dbd = false;
    bool do_raw = false;
    bool force = false;
    bool got_contents = false;
    bool got_mask = false;
    bool mode_6 = false;        /* so default is mode_10 */
    bool rtd = false;   /* added in spc5r11 */
    bool save = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c, num, alloc_len, off, pdt, k, md_len, hdr_len, bd_len;
    int mask_in_len;
    int sg_fd = -1;
    int pg_code = -1;
    int sub_pg_code = 0;
    int verbose = 0;
    int read_in_len = 0;
    int ret = 0;
    unsigned u, uu;
    const char * device_name = NULL;
    const char * cfile_arg = NULL;
    const char * contents_arg = NULL;
    uint8_t read_in[MX_ALLOC_LEN];
    uint8_t mask_in[MX_ALLOC_LEN];
    uint8_t ref_md[MX_ALLOC_LEN];
    char ebuff[EBUFF_SZ];
    char errStr[128];
    char b[80];
    struct sg_simple_inquiry_resp inq_data;
    static const int read_in_sz = sizeof(read_in);
    static const int mask_in_sz = sizeof(mask_in);
    static const int ref_md_sz = sizeof(ref_md);

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "6c:C:dfhl:m:M:p:rRsvV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case '6':
            mode_6 = true;
            break;
        case 'c':
            contents_arg = optarg;
            break;
        case 'C':
            cfile_arg = optarg;
            break;
        case 'd':
            dbd = true;
            break;
        case 'f':
            force = true;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'l':
            num = sscanf(optarg, "%d", &res);
            if ((1 == num) && ((6 == res) || (10 == res)))
                mode_6 = (6 == res);
            else {
                pr2serr("length (of cdb) must be 6 or 10\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'm':
            memset(mask_in, 0xff, mask_in_sz);
            if (0 != build_mask(optarg, false, do_raw, mask_in, &mask_in_len,
                                mask_in_sz)) {
                pr2serr("bad argument to '--mask'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            got_mask = true;
            break;
        case 'M':
            if (0 != build_mask(optarg, true, do_raw, mask_in, &mask_in_len,
                                mask_in_sz)) {
                pr2serr("bad argument or unable to decode '--mfile'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            got_mask = true;
            break;
        case 'p':
           if (NULL == strchr(optarg, ',')) {
                num = sscanf(optarg, "%x", &u);
                if ((1 != num) || (u > 62)) {
                    pr2serr("Bad hex page code value after '--page' "
                            "switch\n");
                    return SG_LIB_SYNTAX_ERROR;
                }
                pg_code = u;
            } else if (2 == sscanf(optarg, "%x,%x", &u, &uu)) {
                if (uu > 254) {
                    pr2serr("Bad hex sub page code value after '--page' "
                            "switch\n");
                    return SG_LIB_SYNTAX_ERROR;
                }
                pg_code = u;
                sub_pg_code = uu;
            } else {
                pr2serr("Bad hex page code, subpage code sequence after "
                        "'--page' switch\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'r':
            do_raw = true;
            break;
        case 'R':
            rtd = true;
            break;
        case 's':
            save = true;
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n", argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (cfile_arg || contents_arg) {
        const char * inp = (cfile_arg ? cfile_arg : contents_arg);

        if (cfile_arg && contents_arg) {
            pr2serr("Cannot have both --contents= and --cfile= options\n");
            return SG_LIB_SYNTAX_ERROR;
        }
        memset(read_in, 0, read_in_sz);
        if ((ret = build_mode_page(inp, !! cfile_arg, do_raw,
                                   read_in, &read_in_len, read_in_sz))) {
            pr2serr("bad argument to '%s'\n", cfile_arg ? "--cfile=" :
                                                          "--contents=");
            return ret;
        }
        if (verbose > 5) {
            pr2serr("Decoded contents:\n");
            hex2stderr(read_in, read_in_len, 1);
        }
        got_contents = true;
    }

#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        /* verbose_given = false; */
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr(ME "version: %s\n", version_str);
        return 0;
    }

    if (NULL == device_name) {
        pr2serr("missing device name!\n\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    if ((pg_code < 0) && (! rtd)) {
        pr2serr("need page code (see '--page=')\n\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }
    if (got_mask && force) {
        pr2serr("cannot use both '--force' and '--mask'\n\n");
        usage();
        return SG_LIB_CONTRADICT;
    }

    sg_fd = sg_cmds_open_device(device_name, false /* rw */, verbose);
    if (sg_fd < 0) {
        if (verbose)
            pr2serr(ME "open error: %s: %s\n", device_name,
                    safe_strerror(-sg_fd));
        ret = sg_convert_errno(-sg_fd);
        goto fini;
    }
    if (rtd)
        goto revert_to_defaults;

    if (0 == sg_simple_inquiry(sg_fd, &inq_data, false, verbose))
        pdt = inq_data.peripheral_type;
    else
        pdt = PDT_UNKNOWN;

    /* do MODE SENSE to fetch current values */
    memset(ref_md, 0, ref_md_sz);
    snprintf(errStr, sizeof(errStr), "MODE SENSE (%d): ", mode_6 ? 6 : 10);
    alloc_len = mode_6 ? SHORT_ALLOC_LEN : MX_ALLOC_LEN;
    if (mode_6)
        res = sg_ll_mode_sense6(sg_fd, dbd, 0 /*current */, pg_code,
                                sub_pg_code, ref_md, alloc_len, true,
                                verbose);
     else
        res = sg_ll_mode_sense10(sg_fd, false /* llbaa */, dbd,
                                 0 /* current */, pg_code, sub_pg_code,
                                 ref_md, alloc_len, true, verbose);
    ret = res;
    if (res) {
        if (SG_LIB_CAT_INVALID_OP == res)
            pr2serr("%snot supported, try '--len=%d'\n", errStr,
                    (mode_6 ? 10 : 6));
        else {
            sg_get_category_sense_str(res, sizeof(b), b, verbose);
            pr2serr("%s%s\n", errStr, b);
        }
        goto fini;
    }
    off = sg_mode_page_offset(ref_md, alloc_len, mode_6, ebuff, EBUFF_SZ);
    if (off < 0) {
        pr2serr("%s%s\n", errStr, ebuff);
        goto fini;
    }
    md_len = sg_msense_calc_length(ref_md, alloc_len, mode_6, &bd_len);
    if (md_len < 0) {
        pr2serr("%ssg_msense_calc_length() failed\n", errStr);
        goto fini;
    }
    hdr_len = mode_6 ? 4 : 8;
    if (got_contents) {
        if (read_in_len < 2) {
            pr2serr("contents length=%d too short\n", read_in_len);
            goto fini;
        }
        ref_md[0] = 0;  /* mode data length reserved for mode select */
        if (! mode_6)
            ref_md[1] = 0;    /* mode data length reserved for mode select */
        if (0 == pdt)   /* for disks mask out DPOFUA bit */
            ref_md[mode_6 ? 2 : 3] &= 0xef;
        if (md_len > alloc_len) {
            pr2serr("mode data length=%d exceeds allocation length=%d\n",
                    md_len, alloc_len);
            goto fini;
        }
        if (got_mask) {
            for (k = 0; k < (md_len - off); ++k) {
                if ((0x0 == mask_in[k]) || (k > read_in_len))
                   read_in[k] = ref_md[off + k];
                else if (mask_in[k] < 0xff) {
                   c = (ref_md[off + k] & (0xff & ~mask_in[k]));
                   read_in[k] = (c | (read_in[k] & mask_in[k]));
                }
            }
            read_in_len = md_len - off;
        }
        if (! force) {
            if ((! (ref_md[off] & 0x80)) && save) {
                pr2serr("PS bit in existing mode page indicates that it is "
                        "not saveable\n    but '--save' option given\n");
                goto fini;
            }
            read_in[0] &= 0x7f; /* mask out PS bit, reserved in mode select */
            if ((md_len - off) != read_in_len) {
                pr2serr("contents length=%d but reference mode page "
                        "length=%d\n", read_in_len, md_len - off);
                goto fini;
            }
            if (pg_code != (read_in[0] & 0x3f)) {
                pr2serr("contents page_code=0x%x but reference "
                        "page_code=0x%x\n", (read_in[0] & 0x3f), pg_code);
                goto fini;
            }
            if ((read_in[0] & 0x40) != (ref_md[off] & 0x40)) {
                pr2serr("contents flags subpage but reference page does not "
                        "(or vice versa)\n");
                goto fini;
            }
            if ((read_in[0] & 0x40) && (read_in[1] != sub_pg_code)) {
                pr2serr("contents subpage_code=0x%x but reference "
                        "sub_page_code=0x%x\n", read_in[1], sub_pg_code);
                goto fini;
            }
        } else
            md_len = off + read_in_len; /* force length */

        memcpy(ref_md + off, read_in, read_in_len);
        if (mode_6)
            res = sg_ll_mode_select6_v2(sg_fd, true /* PF */, rtd, save,
                                        ref_md, md_len, true, verbose);
        else
            res = sg_ll_mode_select10_v2(sg_fd, true /* PF */, rtd, save,
                                         ref_md, md_len, true, verbose);
        ret = res;
        if (res)
            goto fini;
    } else {
        printf(">>> No contents given, so show current mode page data:\n");
        printf("  header:\n");
        hex2stdout(ref_md, hdr_len, -1);
        if (bd_len) {
            printf("  block descriptor(s):\n");
            hex2stdout(ref_md + hdr_len, bd_len, -1);
        } else
            printf("  << no block descriptors >>\n");
        printf("  mode page:\n");
        hex2stdout(ref_md + off, md_len - off, -1);
    }
    ret = 0;
    goto fini;

revert_to_defaults:
    if (verbose)
        pr2serr("Doing MODE SELECT(%d) with revert to defaults (RTD) set "
                "and SP=%d\n", mode_6 ? 6 : 10, !! save);
    if (mode_6)
        res = sg_ll_mode_select6_v2(sg_fd, false /* PF */, true /* rtd */,
                                    save, NULL, 0, true, verbose);
    else
        res = sg_ll_mode_select10_v2(sg_fd, false /* PF */, true /* rtd */,
                                     save, NULL, 0, true, verbose);
    ret = res;
fini:
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if (0 == verbose) {
        if (! sg_if_can2stderr("sg_wr_mode failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' or '-vv' for "
                    "more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
